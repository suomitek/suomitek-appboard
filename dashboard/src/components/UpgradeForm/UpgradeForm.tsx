import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import { JSONSchema4 } from "json-schema";
import * as React from "react";
import * as YAML from "yaml";

import { deleteValue, setValue } from "../../shared/schema";
import { IChartState, IChartVersion } from "../../shared/types";
import * as url from "../../shared/url";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";

export interface IUpgradeFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  chartName: string;
  chartsIsFetching: boolean;
  namespace: string;
  cluster: string;
  releaseName: string;
  repo: string;
  repoNamespace: string;
  error?: Error;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    cluster: string,
    namespace: string,
    version: IChartVersion,
    chartNamespace: string,
    releaseName: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchChartVersions: (namespace: string, id: string) => Promise<IChartVersion[]>;
  getChartVersion: (namespace: string, id: string, chartVersion: string) => void;
}

interface IUpgradeFormState {
  appValues: string;
  valuesModified: boolean;
  isDeploying: boolean;
  modifications?: jsonpatch.Operation[];
  deployedValues?: string;
}

class UpgradeForm extends React.Component<IUpgradeFormProps, IUpgradeFormState> {
  public state: IUpgradeFormState = {
    appValues: this.props.appCurrentValues || "",
    isDeploying: false,
    valuesModified: false,
  };

  public componentDidMount() {
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    this.props.fetchChartVersions(this.props.repoNamespace, chartID);
  }

  public componentDidUpdate = (prevProps: IUpgradeFormProps) => {
    let modifications = this.state.modifications;
    // applyModifications is an expensive operatior, that's why it's only defined within
    // the componentDidUpdate scope
    const applyModifications = (mods: jsonpatch.Operation[], appValues: string) => {
      // And we add any possible change made to the original version
      if (mods.length) {
        mods.forEach(modification => {
          if (modification.op === "remove") {
            appValues = deleteValue(appValues, modification.path);
          } else {
            // Transform the modification as a ReplaceOperation to read its value
            const value = (modification as jsonpatch.ReplaceOperation<any>).value;
            appValues = setValue(appValues, modification.path, value);
          }
        });
      }
      return appValues;
    };

    if (this.props.deployed.values && !modifications) {
      // Calculate modifications from the default values
      const defaultValuesObj = YAML.parse(this.props.deployed.values);
      const deployedValuesObj = YAML.parse(this.props.appCurrentValues || "");
      modifications = jsonpatch.compare(defaultValuesObj, deployedValuesObj);
      const values = applyModifications(modifications, this.props.deployed.values);
      this.setState({ modifications });
      this.setState({ appValues: values, deployedValues: values });
    }

    if (prevProps.selected.version !== this.props.selected.version && !this.state.valuesModified) {
      // Apply modifications to the new selected version
      const appValues = modifications
        ? applyModifications(modifications, this.props.selected.values || "")
        : this.props.selected.values || "";
      this.setState({ appValues });
    }
  };

  public render() {
    const { namespace, releaseName, error, selected } = this.props;
    if (error) {
      return (
        <ErrorSelector error={error} namespace={namespace} action="update" resource={releaseName} />
      );
    }
    if (selected.versions.length === 0) {
      return <LoadingWrapper />;
    }
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    return (
      <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
        <div className="row">
          <div className="col-12">
            <h2>{`${this.props.releaseName} (${chartID})`}</h2>
          </div>
          <div className="col-8">
            <DeploymentFormBody
              deploymentEvent="upgrade"
              chartNamespace={this.props.repoNamespace}
              chartID={chartID}
              chartVersion={this.props.appCurrentVersion}
              deployedValues={this.state.deployedValues}
              chartsIsFetching={this.props.chartsIsFetching}
              namespace={this.props.namespace}
              cluster={this.props.cluster}
              releaseVersion={this.props.appCurrentVersion}
              selected={this.props.selected}
              push={this.props.push}
              goBack={this.props.goBack}
              getChartVersion={this.props.getChartVersion}
              setValues={this.handleValuesChange}
              appValues={this.state.appValues}
              setValuesModified={this.setValuesModified}
            />
          </div>
        </div>
      </form>
    );
  }

  public setValuesModified = () => {
    this.setState({ valuesModified: true });
  };

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value });
  };

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const {
      selected,
      push,
      upgradeApp,
      releaseName,
      cluster,
      namespace,
      repoNamespace,
    } = this.props;
    const { appValues } = this.state;

    this.setState({ isDeploying: true });
    if (selected.version) {
      const deployed = await upgradeApp(
        cluster,
        namespace,
        selected.version,
        repoNamespace,
        releaseName,
        appValues,
        selected.schema,
      );
      this.setState({ isDeploying: false });
      if (deployed) {
        push(url.app.apps.get(cluster, namespace, releaseName));
      }
    }
  };
}

export default UpgradeForm;
