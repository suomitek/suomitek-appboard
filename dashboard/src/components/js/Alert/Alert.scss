.alert {
  &:not(.alert-app-level) {
    .alert-action {
      padding: 0;
      background: transparent;
      border: 0;
      font-size: 13px;
      margin: 0;
      text-transform: none;
      height: auto;
      letter-spacing: normal;
      line-height: 1.15;
      font-weight: 400;

      &:hover {
        background-color: transparent;
      }
    }
  }
}
