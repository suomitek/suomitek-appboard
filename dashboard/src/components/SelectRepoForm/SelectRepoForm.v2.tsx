import { get } from "lodash";
import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";

import Alert from "components/js/Alert";
import * as url from "shared/url";
import { IAppRepository } from "../../shared/types";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper.v2";
import "./SelectRepoForm.css";

interface ISelectRepoFormProps {
  isFetching: boolean;
  cluster: string;
  namespace: string;
  repoError?: Error;
  error?: Error;
  repo: IAppRepository;
  repos: IAppRepository[];
  chartName: string;
  checkChart: (namespace: string, repo: string, chartName: string) => any;
  fetchRepositories: (namespace: string) => void;
}

function SelectRepoForm({
  isFetching,
  cluster,
  namespace,
  repoError,
  error,
  repo,
  repos,
  chartName,
  checkChart,
  fetchRepositories,
}: ISelectRepoFormProps) {
  const [repoName, setRepoName] = useState(get(repo, "metadata.name", ""));

  useEffect(() => {
    fetchRepositories(namespace);
  }, [fetchRepositories, namespace]);

  const handleChartRepoNameChange = async (e: React.ChangeEvent<HTMLSelectElement>) => {
    checkChart(namespace, e.target.value, chartName);
    setRepoName(e.currentTarget.value);
  };

  const getRepoURL = (name: string) => {
    let res = "";
    repos.forEach(r => {
      if (r.metadata.name === name && r.spec) {
        res = r.spec.url;
      }
    });
    return res;
  };

  return (
    <LoadingWrapper loaded={!isFetching}>
      {repoError && <Alert theme="danger">Found error: {repoError.message}</Alert>}
      {!repoError && repos.length === 0 && (
        <Alert theme="warning">
          <h5>Chart repositories not found.</h5>
          Manage your Helm chart repositories in Suomitek-appboard by visiting the{" "}
          <Link to={url.app.config.apprepositories(cluster, namespace)}>
            App repositories configuration
          </Link>{" "}
          page.
        </Alert>
      )}
      {repos.length > 0 && (
        <div className="select-repo-form">
          {error && <Alert theme="danger">Found error: {error.message}</Alert>}
          <h2>Select the source repository of {chartName}</h2>
          <label className="select-repo-form-label" htmlFor="chartRepoName">
            Chart Repository Name *
          </label>
          <div className="clr-select-wrapper">
            <select
              id="chartRepoName"
              onChange={handleChartRepoNameChange}
              value={repoName}
              required={true}
              className="clr-page-size-select"
            >
              {!repoName && <option key="" value="" />}
              {repos.map(r => (
                <option key={r.metadata.name} value={r.metadata.name}>
                  {r.metadata.name} ({getRepoURL(r.metadata.name)})
                </option>
              ))}
            </select>
          </div>
          <p>
            {" "}
            * If the repository containing {chartName} is not in the list add it{" "}
            <Link to={url.app.config.apprepositories(cluster, namespace)}>here</Link>.{" "}
          </p>
        </div>
      )}
    </LoadingWrapper>
  );
}

export default SelectRepoForm;
