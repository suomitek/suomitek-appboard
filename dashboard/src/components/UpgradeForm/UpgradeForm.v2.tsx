import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import { JSONSchema4 } from "json-schema";
import React, { useEffect, useState } from "react";
import * as YAML from "yaml";

import ChartSummary from "components/Catalog/ChartSummary";
import ChartHeader from "components/ChartView/ChartHeader.v2";
import Alert from "components/js/Alert";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { deleteValue, setValue } from "../../shared/schema";
import { IChartState, IChartVersion } from "../../shared/types";
import * as url from "../../shared/url";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody.v2";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper.v2";

export interface IUpgradeFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  chartName: string;
  chartsIsFetching: boolean;
  namespace: string;
  cluster: string;
  releaseName: string;
  repo: string;
  repoNamespace: string;
  error?: Error;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    cluster: string,
    namespace: string,
    version: IChartVersion,
    chartNamespace: string,
    releaseName: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  fetchChartVersions: (namespace: string, id: string) => Promise<IChartVersion[]>;
  getChartVersion: (namespace: string, id: string, chartVersion: string) => void;
}

function applyModifications(mods: jsonpatch.Operation[], values: string) {
  // And we add any possible change made to the original version
  if (mods.length) {
    mods.forEach(modification => {
      if (modification.op === "remove") {
        values = deleteValue(values, modification.path);
      } else {
        // Transform the modification as a ReplaceOperation to read its value
        const value = (modification as jsonpatch.ReplaceOperation<any>).value;
        values = setValue(values, modification.path, value);
      }
    });
  }
  return values;
}

function UpgradeForm({
  appCurrentVersion,
  appCurrentValues,
  chartName,
  chartsIsFetching,
  namespace,
  cluster,
  releaseName,
  repo,
  repoNamespace,
  error,
  selected,
  deployed,
  upgradeApp,
  push,
  fetchChartVersions,
  getChartVersion,
}: IUpgradeFormProps) {
  const [appValues, setAppValues] = useState(appCurrentValues || "");
  const [isDeploying, setIsDeploying] = useState(false);
  const [valuesModified, setValuesModified] = useState(false);
  const [modifications, setModifications] = useState(
    undefined as undefined | jsonpatch.Operation[],
  );
  const [deployedValues, setDeployedValues] = useState("");

  const chartID = `${repo}/${chartName}`;
  const { version } = selected;

  useEffect(() => {
    fetchChartVersions(repoNamespace, chartID);
  }, [fetchChartVersions, repoNamespace, chartID]);

  useEffect(() => {
    if (deployed.values && !modifications) {
      // Calculate modifications from the default values
      const defaultValuesObj = YAML.parse(deployed.values);
      const deployedValuesObj = YAML.parse(appCurrentValues || "");
      const newModifications = jsonpatch.compare(defaultValuesObj, deployedValuesObj);
      const values = applyModifications(newModifications, deployed.values);
      setModifications(newModifications);
      setAppValues(values);
      setDeployedValues(values);
    }
  }, [deployed.values, appCurrentValues, modifications]);

  useEffect(() => {
    if (deployed.chartVersion?.attributes.version) {
      getChartVersion(repoNamespace, chartID, deployed.chartVersion?.attributes.version);
    }
  }, [getChartVersion, repoNamespace, chartID, deployed.chartVersion]);

  useEffect(() => {
    if (!valuesModified && selected.values) {
      // Apply modifications to the new selected version
      const newAppValues = modifications?.length
        ? applyModifications(modifications, selected.values)
        : selected.values;
      setAppValues(newAppValues);
    }
  }, [selected.values, modifications, valuesModified]);

  const setValuesModifiedTrue = () => {
    setValuesModified(true);
  };

  const handleValuesChange = (value: string) => {
    setAppValues(value);
  };

  const selectVersion = (e: React.ChangeEvent<HTMLSelectElement>) => {
    getChartVersion(repoNamespace, chartID, e.currentTarget.value);
  };

  const handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setIsDeploying(true);
    if (selected.version) {
      const deployedSuccess = await upgradeApp(
        cluster,
        namespace,
        selected.version,
        repoNamespace,
        releaseName,
        appValues,
        selected.schema,
      );
      setIsDeploying(false);
      if (deployedSuccess) {
        push(url.app.apps.get(cluster, namespace, releaseName));
      }
    }
  };

  if (error) {
    return <Alert theme="danger">Found error: {error.message}</Alert>;
  }
  if (selected.versions.length === 0 || !version) {
    return <LoadingWrapper loaded={false} />;
  }

  const chartAttrs = version.relationships.chart.data;
  return (
    <section>
      <ChartHeader
        releaseName={releaseName}
        chartAttrs={chartAttrs}
        versions={selected.versions}
        onSelect={selectVersion}
        currentVersion={deployed.chartVersion?.attributes.version}
      />
      {isDeploying && (
        <h3 className="center" style={{ marginBottom: "1.2rem" }}>
          Hang tight, the application is being deployed...
        </h3>
      )}
      <LoadingWrapper loaded={!isDeploying}>
        <Row>
          <Column span={3}>
            <ChartSummary version={version} chartAttrs={chartAttrs} />
          </Column>
          <Column span={9}>
            <form onSubmit={handleDeploy}>
              <DeploymentFormBody
                deploymentEvent="upgrade"
                chartID={chartID}
                chartVersion={appCurrentVersion}
                deployedValues={deployedValues}
                chartsIsFetching={chartsIsFetching}
                selected={selected}
                setValues={handleValuesChange}
                appValues={appValues}
                setValuesModified={setValuesModifiedTrue}
              />
            </form>
          </Column>
        </Row>
      </LoadingWrapper>
    </section>
  );
}

export default UpgradeForm;
