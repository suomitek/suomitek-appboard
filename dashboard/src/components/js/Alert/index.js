import Alert, { AlertThemes } from "./Alert";
export { AlertThemes };
export default Alert;
