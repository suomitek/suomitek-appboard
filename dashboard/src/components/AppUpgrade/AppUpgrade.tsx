import * as React from "react";

import { RouterAction } from "connected-react-router";
import { JSONSchema4 } from "json-schema";
import { IAppRepository, IChartState, IChartVersion, IRelease } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";

export interface IAppUpgradeProps {
  app?: IRelease;
  appsIsFetching: boolean;
  chartsIsFetching: boolean;
  appsError: Error | undefined;
  namespace: string;
  cluster: string;
  releaseName: string;
  repoName?: string;
  repoNamespace?: string;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    cluster: string,
    namespace: string,
    version: IChartVersion,
    chartNamespace: string,
    releaseName: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  fetchChartVersions: (namespace: string, id: string) => Promise<IChartVersion[]>;
  getAppWithUpdateInfo: (cluster: string, namespace: string, releaseName: string) => void;
  getChartVersion: (namespace: string, id: string, chartVersion: string) => void;
  getDeployedChartVersion: (namespace: string, id: string, chartVersion: string) => void;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  // repo selector properties
  reposIsFetching: boolean;
  kubeappsNamespace: string;
  repoError?: Error;
  chartsError: Error | undefined;
  repo: IAppRepository;
  repos: IAppRepository[];
  checkChart: (namespace: string, repo: string, chartName: string) => any;
  fetchRepositories: (namespace: string) => void;
}

class AppUpgrade extends React.Component<IAppUpgradeProps> {
  public componentDidMount() {
    const { releaseName, getAppWithUpdateInfo, cluster, namespace } = this.props;
    getAppWithUpdateInfo(cluster, namespace, releaseName);
  }

  public componentDidUpdate(prevProps: IAppUpgradeProps) {
    const { app, repoName, repoNamespace } = this.props;
    if (app && repoName && repoNamespace) {
      const { chart } = app;
      if (
        chart &&
        chart.metadata &&
        chart.metadata.name &&
        chart.metadata.version &&
        (prevProps.app !== app || prevProps.repoName !== repoName)
      ) {
        const chartID = `${repoName}/${chart.metadata.name}`;
        this.props.getDeployedChartVersion(repoNamespace, chartID, chart.metadata.version);
      }
    }
  }

  public render() {
    const {
      app,
      namespace,
      cluster,
      appsError,
      releaseName,
      appsIsFetching,
      chartsIsFetching,
      repoName,
      repoNamespace,
      selected,
      deployed,
      upgradeApp,
      push,
      goBack,
      fetchChartVersions,
      getChartVersion,
    } = this.props;
    if (appsError) {
      return (
        <ErrorSelector
          error={appsError}
          namespace={namespace}
          action="update"
          resource={releaseName}
        />
      );
    }
    if (appsIsFetching || !app || !app.updateInfo) {
      return <LoadingWrapper />;
    }
    const repo = repoName || app.updateInfo.repository.name;
    const repoNS = repoNamespace || app.updateInfo.repository.namespace;
    if (app && app.chart && app.chart.metadata && repo) {
      return (
        <div>
          <UpgradeForm
            appCurrentVersion={app.chart.metadata.version!}
            appCurrentValues={(app.config && app.config.raw) || ""}
            chartName={app.chart.metadata.name!}
            chartsIsFetching={chartsIsFetching}
            repo={repo}
            repoNamespace={repoNS}
            namespace={namespace}
            cluster={cluster}
            releaseName={releaseName}
            selected={selected}
            deployed={deployed}
            upgradeApp={upgradeApp}
            push={push}
            goBack={goBack}
            fetchChartVersions={fetchChartVersions}
            getChartVersion={getChartVersion}
          />
        </div>
      );
    }

    return (
      <SelectRepoForm
        isFetching={this.props.reposIsFetching}
        error={this.props.chartsError}
        kubeappsNamespace={this.props.kubeappsNamespace}
        cluster={this.props.cluster}
        namespace={this.props.namespace}
        repoError={this.props.repoError}
        repo={this.props.repo}
        repos={this.props.repos}
        chartName={app.chart?.metadata?.name!}
        checkChart={this.props.checkChart}
        fetchRepositories={this.props.fetchRepositories}
      />
    );
  }
}

export default AppUpgrade;
