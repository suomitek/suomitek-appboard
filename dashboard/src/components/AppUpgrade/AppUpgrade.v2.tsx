import React, { useEffect } from "react";

import Alert from "components/js/Alert";
import { RouterAction } from "connected-react-router";
import { JSONSchema4 } from "json-schema";
import { IAppRepository, IChartState, IChartVersion, IRelease } from "../../shared/types";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper.v2";
import SelectRepoForm from "../SelectRepoForm/SelectRepoForm.v2";
import UpgradeForm from "../UpgradeForm/UpgradeForm.v2";

export interface IAppUpgradeProps {
  app?: IRelease;
  appsIsFetching: boolean;
  chartsIsFetching: boolean;
  appsError: Error | undefined;
  namespace: string;
  cluster: string;
  releaseName: string;
  repoName?: string;
  repoNamespace?: string;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    cluster: string,
    namespace: string,
    version: IChartVersion,
    chartNamespace: string,
    releaseName: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  fetchChartVersions: (namespace: string, id: string) => Promise<IChartVersion[]>;
  getAppWithUpdateInfo: (cluster: string, namespace: string, releaseName: string) => void;
  getChartVersion: (namespace: string, id: string, chartVersion: string) => void;
  getDeployedChartVersion: (namespace: string, id: string, chartVersion: string) => void;
  push: (location: string) => RouterAction;
  // repo selector properties
  reposIsFetching: boolean;
  repoError?: Error;
  chartsError: Error | undefined;
  repo: IAppRepository;
  repos: IAppRepository[];
  checkChart: (namespace: string, repo: string, chartName: string) => any;
  fetchRepositories: (namespace: string) => void;
}

function AppUpgrade({
  app,
  appsIsFetching,
  chartsIsFetching,
  appsError,
  namespace,
  cluster,
  releaseName,
  repoName,
  repoNamespace,
  selected,
  deployed,
  upgradeApp,
  fetchChartVersions,
  getAppWithUpdateInfo,
  getChartVersion,
  getDeployedChartVersion,
  push,
  reposIsFetching,
  repoError,
  chartsError,
  repo,
  repos,
  checkChart,
  fetchRepositories,
}: IAppUpgradeProps) {
  useEffect(() => {
    getAppWithUpdateInfo(cluster, namespace, releaseName);
  }, [getAppWithUpdateInfo, cluster, namespace, releaseName]);

  const chart = app?.chart;
  useEffect(() => {
    if (
      repoName &&
      repoNamespace &&
      chart &&
      chart.metadata &&
      chart.metadata &&
      chart.metadata.name &&
      chart.metadata.version
    ) {
      const chartID = `${repoName}/${chart.metadata.name}`;
      getDeployedChartVersion(repoNamespace, chartID, chart.metadata.version);
    }
  }, [getDeployedChartVersion, app, chart, repoName, repoNamespace]);

  if (appsError) {
    return <Alert theme="danger">Found error: {appsError.message}</Alert>;
  }
  if (appsIsFetching || !app || !app.updateInfo) {
    return <LoadingWrapper loaded={false} />;
  }

  const appRepoName = repoName || app.updateInfo.repository.name;
  const repoNS = repoNamespace || app.updateInfo.repository.namespace;
  if (app && app.chart && app.chart.metadata && appRepoName) {
    return (
      <div>
        <UpgradeForm
          appCurrentVersion={app.chart.metadata.version!}
          appCurrentValues={(app.config && app.config.raw) || ""}
          chartName={app.chart.metadata.name!}
          chartsIsFetching={chartsIsFetching}
          repo={appRepoName}
          repoNamespace={repoNS}
          namespace={namespace}
          cluster={cluster}
          releaseName={releaseName}
          selected={selected}
          deployed={deployed}
          upgradeApp={upgradeApp}
          push={push}
          fetchChartVersions={fetchChartVersions}
          getChartVersion={getChartVersion}
        />
      </div>
    );
  }

  return (
    <SelectRepoForm
      isFetching={reposIsFetching}
      cluster={cluster}
      error={chartsError}
      namespace={namespace}
      repoError={repoError}
      repo={repo}
      repos={repos}
      chartName={chart?.metadata?.name!}
      checkChart={checkChart}
      fetchRepositories={fetchRepositories}
    />
  );
}

export default AppUpgrade;
