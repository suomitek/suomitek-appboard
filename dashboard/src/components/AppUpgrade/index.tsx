import * as React from "react";

import { IAppUpgradeProps } from "./AppUpgrade";

const AppUpgrade = React.lazy(() => import("./AppUpgrade"));
const AppUpgradeV2 = React.lazy(() => import("./AppUpgrade.v2"));

interface IAppUpgradeSelectorProps extends IAppUpgradeProps {
  UI: string;
}

const AppUpgradeSelector: React.FC<IAppUpgradeSelectorProps> = props => (
  <React.Suspense fallback={null}>
    {props.UI === "clarity" ? <AppUpgradeV2 {...props} /> : <AppUpgrade {...props} />}
  </React.Suspense>
);

export default AppUpgradeSelector;
