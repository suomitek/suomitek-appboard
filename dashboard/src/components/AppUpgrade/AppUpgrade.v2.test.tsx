import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import Alert from "components/js/Alert";
import { hapi } from "shared/hapi/release";
import { defaultStore, mountWrapper } from "shared/specs/mountWrapper";
import { IAppRepository, IChartState, IChartVersion, IRelease } from "shared/types";
import itBehavesLike from "../../shared/specs";
import SelectRepoForm from "../SelectRepoForm/SelectRepoForm.v2";
import UpgradeForm from "../UpgradeForm/UpgradeForm.v2";
import AppUpgrade, { IAppUpgradeProps } from "./AppUpgrade.v2";

const versions = [
  {
    id: "foo",
    attributes: { version: "1.2.3" },
    relationships: { chart: { data: { repo: { name: "bitnami" } } } },
  },
  {
    id: "foo",
    attributes: { version: "1.2.4" },
    relationships: { chart: { data: { repo: { name: "bitnami" } } } },
  },
] as IChartVersion[];
const schema = { properties: { foo: { type: "string" } } };

const defaultProps = {
  app: {} as hapi.release.Release,
  appsIsFetching: false,
  chartsIsFetching: false,
  reposIsFetching: false,
  repoName: "",
  repoNamespace: "chart-namespace",
  isFetching: false,
  checkChart: jest.fn(),
  clearRepo: jest.fn(),
  appsError: undefined,
  chartsError: undefined,
  fetchChartVersions: jest.fn(),
  fetchRepositories: jest.fn(),
  getAppWithUpdateInfo: jest.fn(),
  getChartVersion: jest.fn(),
  deployed: {} as IChartState["deployed"],
  getDeployedChartVersion: jest.fn(),
  kubeappsNamespace: "kubeapps",
  namespace: "default",
  cluster: "default",
  push: jest.fn(),
  goBack: jest.fn(),
  releaseName: "foo",
  repo: {} as IAppRepository,
  repoError: undefined,
  repos: [],
  selected: { versions, version: versions[0], schema },
  upgradeApp: jest.fn(),
  version: "1.0.0",
} as IAppUpgradeProps;

beforeEach(() => {
  jest.resetAllMocks();
});

itBehavesLike("aLoadingComponent", {
  component: AppUpgrade,
  props: { ...defaultProps, isFetching: true },
});

it("renders the repo selection form if not introduced", () => {
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
          updateInfo: { repository: {} },
        } as IRelease
      }
      repos={[
        {
          metadata: { name: "stable" },
        } as IAppRepository,
      ]}
    />,
  );
  expect(wrapper.find(SelectRepoForm)).toExist();
  expect(wrapper.find(Alert)).not.toExist();
  expect(wrapper.find(UpgradeForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

context("when an error exists", () => {
  it("renders a generic error message", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const wrapper = shallow(
      <AppUpgrade
        {...defaultProps}
        appsError={new Error("foo does not exist")}
        repos={[repo]}
        repo={repo}
      />,
    );

    expect(wrapper.find(Alert)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.html()).toContain("foo does not exist");
  });

  it("renders a warning message if there are no repositories", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <AppUpgrade
        {...defaultProps}
        app={
          {
            chart: {
              metadata: {
                name: "bar",
                version: "1.0.0",
              },
            },
            name: "foo",
            updateInfo: { repository: {} },
          } as IRelease
        }
        repos={[]}
      />,
    );

    expect(wrapper.find(SelectRepoForm).find(Alert)).toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(
      wrapper
        .find(Alert)
        .children()
        .text(),
    ).toContain("Chart repositories not found");
  });
});

it("renders the upgrade form when the repo is available", () => {
  const wrapper = mountWrapper(
    defaultStore,
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
          updateInfo: { repository: {} },
        } as IRelease
      }
      repoName="foobar"
    />,
  );
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(Alert)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
});

it("skips the repo selection form if the app contains upgrade info", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const app = {
    chart: {
      metadata: {
        name: "bar",
        version: "1.0.0",
      },
    },
    name: "foo",
    updateInfo: {
      upToDate: true,
      chartLatestVersion: "1.1.0",
      appLatestVersion: "1.1.0",
      repository: { name: "stable", url: "" },
    },
  } as IRelease;
  const wrapper = mountWrapper(
    defaultStore,
    <AppUpgrade {...defaultProps} repos={[repo]} app={app} />,
  );
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(Alert)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
});

describe("when receiving new props", () => {
  it("should request the deployed chart when the app and repo are populated", () => {
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    mountWrapper(
      defaultStore,
      <AppUpgrade
        {...defaultProps}
        getDeployedChartVersion={getDeployedChartVersion}
        repoName="stable"
        app={app}
      />,
    );
    expect(getDeployedChartVersion).toHaveBeenCalledWith(
      defaultProps.repoNamespace,
      "stable/bar",
      "1.0.0",
    );
  });

  it("should request the deployed chart when the repo is populated later", () => {
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    mountWrapper(
      defaultStore,
      <AppUpgrade
        {...defaultProps}
        app={app}
        getDeployedChartVersion={getDeployedChartVersion}
        repoName="stable"
      />,
    );
    expect(getDeployedChartVersion).toHaveBeenCalledWith(
      defaultProps.repoNamespace,
      "stable/bar",
      "1.0.0",
    );
  });
});
