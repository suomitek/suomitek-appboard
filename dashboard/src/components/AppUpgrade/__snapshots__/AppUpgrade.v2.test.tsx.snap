// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
/>
`;

exports[`renders the repo selection form if not introduced 1`] = `
<SelectRepoForm
  chartName="bar"
  checkChart={[MockFunction]}
  cluster="default"
  fetchRepositories={[MockFunction]}
  isFetching={false}
  namespace="default"
  repo={Object {}}
  repos={
    Array [
      Object {
        "metadata": Object {
          "name": "stable",
        },
      },
    ]
  }
/>
`;
